File: Data Structures & Algorithms/combination-target-sum/submission-2.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> comb = new ArrayList<>();
        Arrays.sort(nums);
        dfs(0,target,nums,res,comb);
        return res;
    }
    private void dfs(int i,int rem,int[] nums,List<List<Integer>> res,List<Integer> comb){
        if(rem == 0){
            res.add(new ArrayList<>(comb));
            return;
        }
        for(int j = i; j < nums.length;j++){
            if (rem - nums[j] < 0){
                return;
            }
            comb.add(nums[j]);
            dfs(j,rem - nums[j],nums,res,comb);
            comb.remove(comb.size() -1);
        }
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        comb = []
        candidates.sort()
        def dfs(i,rem):
            if rem == 0:
                res.append(comb[:])
                return
            if i>= len(candidates) or rem - candidates[i] < 0:
                return
            comb.append(candidates[i])
            dfs(i+1,rem-candidates[i])
            while(i+1 < len(candidates) and  candidates[i] == candidates[i+1]):
                i+=1
            comb.pop()
            dfs(i+1,rem)
        dfs(0,target)
        return res